### PR TITLE
Rebase ISO8601 formatting on top of new Gregorian Calendar implementation

### PR DESCRIPTION
**Pull request by parkera**
Use @itingliu's new Gregorian Calendar implementation for ISO8601 formatting. This means:

1. We can sink ISO8601 formatting into Essentials, as it no longer depends on ICU
2. We can improve performance of formatting by not using ICU
3. We can (in a separate PR) make Date's description use this

I also added new benchmarks for ISO8601 formatting. Here are the results:

*OLD*

```
iso8601
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Throughput (# / s) (K)    │     112 │     109 │     108 │     106 │     104 │      89 │      64 │     321 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    8915 │    9126 │    9216 │    9363 │    9527 │   10420 │   11878 │     321 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (ns) *  │    8929 │    9150 │    9249 │    9404 │    9617 │   11248 │   15611 │     321 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

*NEW*

```
iso8601
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Throughput (# / s) (K)    │     491 │     476 │     467 │     459 │     451 │     413 │     299 │    1000 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    2035 │    2084 │    2099 │    2128 │    2154 │    2279 │    3265 │    1000 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (ns) *  │    2035 │    2101 │    2142 │    2181 │    2216 │    2417 │    3349 │    1000 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

Bottom line: The new implementation is about *5x* faster than the old one, in terms of throughput.

The improvements come from several areas including improvements in the `ISO8601FormatStyle` type itself. I think there is actually room for even more here, but I don't want to refactor the Gregorian Calendar implementation until Tina is done with it.

I also pulled in an (internal-only) implementation of `OutputBuffer` from @glessard's [base64 PR](https://github.com/apple/swift-foundation/pull/337). Using this was worth a 25% improvement by avoiding retains, releases, and exclusivity checking in the formatting part of the code.

### CODE REVIEW

**Comment by parkera** [comment]
@swift-ci test

**Review by glessard — APPROVED** [review]

**Comment by FranzBusch** [comment]
This is amazing! Thanks for all the hard work here.

File: Sources/FoundationEssentials/Formatting/Date+ISO8601FormatStyle.swift [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+            var buffer = OutputBuffer(initializing: _buffer.baseAddress!, capacity: _buffer.count)
+            
+            let asciiZero = CChar(48)
+            
+            func append(_ i: Int, zeroPad: Int, buffer: inout OutputBuffer<CChar>) {
+                if i < 10 {
+                    if zeroPad - 1 > 0 {
+                        for _ in 0..<zeroPad-1 { buffer.appendElement(asciiZero) }
+                    }
+                    buffer.appendElement(CChar(i + 48))
```
Comment (glessard) [inline Sources/FoundationEssentials/Formatting/Date+ISO8601FormatStyle.swift:305]: Maybe we could use the `asciiZero` value instead of `48` here and below

Comment (parkera) [reply Sources/FoundationEssentials/Formatting/Date+ISO8601FormatStyle.swift:305]: Yah, I can fix that in a follow-up PR.

**Pull request merged**